server/channels/app/platform/websocket_reliable.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package platform
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+func (ps *PlatformService) GetWSQueues(userID, connectionID string, seqNum int64) (*model.WSQueues, error) {
+	hub := ps.GetHubForUserId(userID)
+	if hub == nil {
+		return nil, nil
+	}
+	connRes := hub.CheckConn(userID, connectionID)
+	if connRes == nil {
+		return nil, nil
+	}
+	aq := connRes.ActiveQueue
+	dq := connRes.DeadQueue
+	dqPtr := connRes.DeadQueuePointer
+
+	// Nothing was written on this server. Early return.
+	if dq[0] == nil {
+		return nil, nil
+	}
+
+	// If seq-1 == last value in the dq.
+	if perfectMatch := !_hasMsgLoss(dq, dqPtr, seqNum); perfectMatch {
+		aqSlice, err := ps.getAQ(aq, connectionID, userID)
+		defer close(aq)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get from active queue: %w", err)
+		}
+		// send only aq
+		return &model.WSQueues{
+			ActiveQ:    aqSlice,
+			ReuseCount: connRes.ReuseCount,
+		}, nil
+	}
+
+	// If seq is there somewhere else in the dq.
+	if ok, index := _isInDeadQueue(dq, seqNum); ok {
+		aqSlice, err := ps.getAQ(aq, connectionID, userID)
+		defer close(aq)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get from active queue: %w", err)
+		}
+		dqSlice, err := ps.getDQ(dq, index, dqPtr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get from dead queue: %w", err)
+		}
+		// send aq + drainedDq.
+		return &model.WSQueues{
+			ActiveQ:    aqSlice,
+			DeadQ:      dqSlice,
+			ReuseCount: connRes.ReuseCount,
+		}, nil
+	}
+
+	// Nothing matched.
+	return nil, nil
+}
+
+func (ps *PlatformService) getAQ(aq <-chan model.WebSocketMessage, connID, userID string) ([]model.ActiveQueueItem, error) {
```

Comment:
Just a naming nit: these two methods are mostly transforming the input data (e.g., channel to slice), not getting it from anywhere, so I was a bit confused at first.